pdp/doc_helpers/__init__.py

```diff
@@ -0,0 +1,172 @@
+by a list item that is a dictionary whose key is the directory name and
+whose value is likewise a list of files or directories, recursively.
+
+For more details, see the README on this subject.
+"""
+from __future__ import print_function
+import subprocess
+from warnings import warn
+import urllib2
+import yaml
+import yaml.scanner
+
+
+def fetch_manifest(base_url, file_name="manifest.yaml"):
+    """
+    Fetch a manifest file and convert it to a Python object.
+
+    :param base_url: URL of external document directory.
+    :param file_name: Name of manifest file.
+    :return: nested lists/dicts
+    """
+    url = "{}/{}".format(base_url, file_name)
+    try:
+        response = urllib2.urlopen(url)
+        try:
+            manifest = yaml.load(response)
+        except yaml.scanner.ScannerError:
+            warn(
+                "User docs: Error parsing manifest file from {}. "
+                "Returning empty manifest".format(url)
+            )
+            manifest = []
+        response.close()
+    except urllib2.URLError:
+        warn(
+            "User docs: Error fetching manifest file from {}. "
+            "Returning empty manifest".format(url)
+        )
+        manifest = []
+    return manifest
+
+
+def flatten_manifest(manifest, path=()):
+    """
+    Generator that converts recursive manifest structure to a sequence of
+    tuples that specify the path to each object (relative to document
+    directory). This makes it convenient to iterate over each file to be
+    downloaded.
+
+    Each tuple contains a sequence of file path components. For example,
+
+        ("images", "abc.png")
+
+    :param manifest: Manifest (see above)
+    :yield: sequence of tuples containing paths to files
+    """
+    if isinstance(manifest, (tuple, list)):
+        for item in manifest:
+            for p in flatten_manifest(item, path):
+                yield p
+    elif isinstance(manifest, dict):
+        for key, item in manifest.items():
+            for p in flatten_manifest(item, path + (key,)):
+                yield p
+    else:  # it's a string
+        yield path + (manifest,)
+
+
+def prep_for_downloads(target_dir):
+    """
+    Prepare for download to target directory from flattened manifest.
+    This basically means removing the target directory.
+
+    :param target_dir:
+    :return:
+    """
+    subprocess.call(["rm", "-rf", target_dir])
+
+
+def download_manifest_item(base_url, manifest_item, target_dir):
+    """
+    Download a file specified by a flattened manifest item,
+    storing it under the specified target directory (it may be in a subdir).
+
+    :param base_url: URL of directory containing files to download.
+    :param manifest_item: Flattened manifest item specifying file to download.
+    :param target_dir: Path specifying top-level target directory in which
+        to store file.
+    """
+    # Make target directory
+    item_dir = "/".join(manifest_item[:-1])
```

Comment:
Advisable to use `os.path.join` throughout, but I guess this branch is stuck in Python 2. This should be flagged for @eyvorchuk 's 2-to-3 conversion.

Reply:
We are indeed in Python 2.

Reply:
No, wait, we have [`os.path.join` in Py 2.](https://docs.python.org/2.7/library/os.path.html#os.path.join)

Reply:
Done